### DISCUSSION TITLE
How far from WF4

### DISCUSSION THREAD

**Discussion by Fabsky** [elsa-workflows/elsa-core#3002] [General]
Hello,

A few months ago I spend several hours to make research to find a replacement to our WF4 engine.  After a lot of time I found Elsa, ans it seems like a gem and really deserve more pub / articles. With the project quality, that's quite strange that it took me so much time to find it (or I need to improve my google skills :D)

So now I've a bit more time I'm analyzing features to see if there is a big gap from WF4.

I'm reading everything I can find since 2 days, it looks like the core features are there and I'll not miss something big.
I think the complicated part for me will be the designer: we've rehosted the WF designer in a WPF app and made a workflow designer that

1. Is decoupled from our back-end server which execute the workflow => Seems OK with Elsa designer
2. Save workflow in XML using our backend api => Seems OK with Elsa designer
3. Allow tracking execution, i.e. activities executed on back-end are shown in the designer to help debugging => Seems not yet available in Elsa
4. Require authentication => I read several post where some say it was possible, other not possible so I don't know ?
5. With activity designer that show data from our back-end (possible because user is authenticated)
6. doesn't need a single line of script/code: everything is done using drag & drop and user don't need any programming skills ==> This is were I think the gap is the biggest

Here are small samples (activities are coming from a list on the left, but I'm sure you already know how the WF4 designer is done):

Control flow:
![2022-05-07_09h15_34](https://user-images.githubusercontent.com/9216452/167243859-e0393d92-f86e-48e7-95f8-517ff7f9ce1a.gif)

Send email using a template (which made something like a mail merge between data & template):
![2022-05-07_09h31_041](https://user-images.githubusercontent.com/9216452/167244172-8d844c76-22d5-4764-93dc-1411fb5fb0bb.gif)

The drop down list all contains data from our back-end, i.e. in the latest sample allowing user to choose which template to use (template has been previously created in our backend), or which property to get value etc.

Do you think it's something we can achieve with Elsa ?

Thank you :
Fabrice

**Comment by sfmskywalker** [comment]
Hi @Fabsky 

Thank you for your questions and the nice words!

1. Indeed, the designer is completely decoupled from the server. The designer is a reusable web component that you can use in your own web application.
2. Elsa stores workflow definitions in JSON format.
3. The dashboard shows a workflow journal of a given workflow instance to help troubleshooting. It's not perfect, but it might be a good start.
4. Although the dashboard doesn't have authentication built-in, the way to do this is by implementing a small JavaScript function that updates outgoing HTTP requests with credentials (e.g. JWT bearer tokens). It's up to your workflow server to then validate these tokens. It's very flexible, allowing you to use whatever authentication mechanism you want.
5. Exactly. One customer did this by programmatically replacing a free-form text field with a dropdown list of URLS that come from their backend. They did this for an **existing** activity (just to brag a little bit about Elsa's flexibility ;) ) and of course when you write custom activities, you have even more control. You can provide available options for e.g. a dropdown list server-side using some provider strategy, or let your custom activity fetch items on the designer using an HTTP request to your backend.
6. You could design all of your activities to not have free-form text fields where the user needs to input expressions, but instead select values from dropdowns, as per your screen recordings.

That being said, I think the biggest difference between Elsa2 and WF4 is that WF4 has native support for activity composition, which works differently in Elsa2. In Elsa2, the flow-chart style diagrams can become unwieldy when the workflow becomes sufficiently complex with low-level activities. But, if you stick to using high level custom activities, it's pretty great.

Elsa3 on the other hand follows WF4's model closely, where you have very similar activities like Sequence, If, ForEach, etc.
It will initially ship with a designer very similar to that of Elsa 1 (with free drag & drop nodes), but the plan is to also ship a designer that is basically what you see in your screen recordings. Having this visual containment helps manage larger workflows.
It's also easier to implement custom activities using composition (where you construct higher-level activities from lower-level ones).

I hope this helps, but let me know if you have nay additional questions!

**Reply by Fabsky** [reply]
I read you post before and forgot to thank you about you reply !
I follow your project and will check again with the V3 ;)
